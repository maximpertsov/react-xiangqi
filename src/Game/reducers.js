import update from 'immutability-helper';
import { useReducer } from 'react';
import XiangqiBoard, { RefType } from '../logic';

// Initial State

// TODO: define in logic class
/* eslint-disable-next-line max-len */
const DEFAULT_FEN = 'rheakaehr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RHEAKAEHR';

const getInitialMovesState = (fen = DEFAULT_FEN) => ([
  {
    piece: undefined,
    fromPos: undefined,
    toPos: undefined,
    board: new XiangqiBoard({ fen }),
  },
]);

const initialPlayers = [
  { name: undefined, color: 'red' },
  { name: undefined, color: 'black' },
];

const getInitialState = () => ({
  loading: true,
  moves: getInitialMovesState(),
  moveCount: 0,
  players: initialPlayers,
  selectedMoveIdx: 0,
});

// Actions

const incrementMoveCount = (state) => (
  { ...state, moveCount: state.moveCount + 1 }
);

const setSelectedMove = (state, index) => (
  { ...state, selectedMoveIdx: index }
);

const selectLastMove = (state) => {
  const { moves } = state;
  return setSelectedMove(state, moves.length - 1);
};

const addMove = (state, board, { fromSlot, toSlot }) => {
  const newMove = {
    fromPos: board.getRankFile(fromSlot),
    toPos: board.getRankFile(toSlot),
    piece: board.getPiece(fromSlot),
    board: board.move(fromSlot, toSlot),
  };
  return selectLastMove(
    incrementMoveCount(
      update(state, { moves: { $push: [newMove] } }),
    ),
  );
};

const setMove = (
  state,
  { piece, origin: fromPos, destination: toPos },
  fromMoveIdx = undefined,
) => {
  const { board } = state.moves[fromMoveIdx || state.moves.length - 1];
  const newMove = {
    piece,
    fromPos,
    toPos,
    board: board.move(fromPos, toPos, RefType.RANK_FILE),
  };
  return incrementMoveCount(update(state, { moves: { $push: [newMove] } }));
};


const setMoves = (state, moves) => selectLastMove(
  moves.reduce(
    (prevState, move) => setMove(prevState, move),
    {
      ...state,
      moves: getInitialMovesState(),
      moveCount: 0,
      setSelectedMove: 0,
      loading: false,
    },
  ),
);

/* eslint-disable-next-line complexity */
const reducer = (state, action) => {
  switch (action.type) {
    case 'add_move':
      return addMove(state, action.board, action.move);
    case 'select_move':
      return setSelectedMove(state, action.index);
    case 'set_moves':
      return setMoves(state, action.moves);
    case 'set_players':
      return ({ ...state, players: action.players });
    default:
      return state;
  }
};

const useGameReducer = () => useReducer(
  reducer, getInitialState(),
);

export default useGameReducer;
