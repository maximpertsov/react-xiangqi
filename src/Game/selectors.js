// Select a move by idx, where -1 is the latest move
// Return an empty object if move lookup fails
export const getMove = ({ moves }, idx) => moves[idx];

export const getNextMoveColor = ({ moves }) => {
  if (moves.length === 0) return 'red';

  // TODO: we don't really need a specific board for this function
  const { piece: lastMovedPiece, board } = moves[moves.length - 1];

  return board.isRedCode(lastMovedPiece) ? 'black' : 'red';
};

const lookupPlayer = (players, key, value) => players
  .find((p) => p[key] === value);

export const getNextMovePlayer = ({ players, moves }) => lookupPlayer(
  players, 'color', getNextMoveColor({ moves }),
);

export default {};
