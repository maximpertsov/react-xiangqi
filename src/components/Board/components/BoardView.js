import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { useSelector } from 'react-redux';

import {
  getBottomPlayerIsRed,
  getNextMoveColor,
  getSelectedMove,
  getTargets,
} from 'reducers';

import * as styles from 'commonStyles';

import Square from './Square';

import boardImg from '../board-1000px.svg.png';

const Wrapper = styled.div`
  background-image: url(${boardImg});
  background-size: contain;
  background-repeat: no-repeat;
  background-position: top;
  display: grid;
  ${styles.MEDIA_TINY} {
    grid-template-rows: repeat(10, ${styles.SQUARE_SIZE_TINY});
    grid-template-columns: repeat(9, ${styles.SQUARE_SIZE_TINY});
  }
  ${styles.MEDIA_SMALL} {
    grid-template-rows: repeat(10, ${styles.SQUARE_SIZE_SMALL});
    grid-template-columns: repeat(9, ${styles.SQUARE_SIZE_SMALL});
  }
  ${styles.MEDIA_MEDIUM} {
    grid-template-rows: repeat(10, ${styles.SQUARE_SIZE_MEDIUM});
    grid-template-columns: repeat(9, ${styles.SQUARE_SIZE_MEDIUM});
  }
  ${styles.MEDIA_LARGE} {
    grid-template-rows: repeat(10, ${styles.SQUARE_SIZE_LARGE});
    grid-template-columns: repeat(9, ${styles.SQUARE_SIZE_LARGE});
  }
`;

const BoardView = ({ handleSquareClick }) => {
  const bottomPlayerIsRed = useSelector(state => getBottomPlayerIsRed(state));
  const nextMoveColor = useSelector(state => getNextMoveColor(state));
  const selectedSlot = useSelector(state => state.selectedSlot);
  const targets = useSelector(state => getTargets(state));
  const {
    board,
    fromSlot: moveFromSlot,
    toSlot: moveToSlot,
  } = useSelector(state => getSelectedMove(state));

  const getPieceCode = useCallback(slot => board.getPiece(slot) || undefined, [
    board,
  ]);

  const getSlot = useCallback(
    (b, i) => (bottomPlayerIsRed ? i : b.length - i - 1),
    [bottomPlayerIsRed],
  );

  const inLastMove = useCallback(
    slot => slot === moveFromSlot || slot === moveToSlot,
    [moveFromSlot, moveToSlot],
  );

  const renderSquares = () =>
    board.board.map((_, i, b) => {
      const slot = getSlot(b, i);
      return (
        <Square
          className="Square"
          key={slot}
          handleClick={handleSquareClick(slot)}
          inCheck={board.inCheck({ slot, nextMoveColor })}
          inLastMove={inLastMove(slot)}
          pieceCode={getPieceCode(slot)}
          selected={selectedSlot === slot}
          targeted={targets.includes(slot)}
        />
      );
    });

  return <Wrapper className="BoardView">{renderSquares()}</Wrapper>;
};

BoardView.propTypes = {
  handleSquareClick: PropTypes.func.isRequired,
};

export default BoardView;
