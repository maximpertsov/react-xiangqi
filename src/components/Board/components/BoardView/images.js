import flatten from 'lodash/flatten';

import bottomLeft from './assets/square-bottom-left.svg';
import bottomRight from './assets/square-bottom-right.svg';
import bottom from './assets/square-bottom.svg';
import inner from './assets/square-inner.svg';
import left from './assets/square-left.svg';
import right from './assets/square-right.svg';
import topLeft from './assets/square-top-left.svg';
import topRight from './assets/square-top-right.svg';
import top from './assets/square-top.svg';

const imageByIndex = Object.freeze([
  [topLeft, top, top, null, null, null, top, top, topRight],
  [left, inner, inner, null, null, null, inner, inner, right],
  [left, inner, inner, null, null, null, inner, inner, right],
  [left, inner, inner, inner, inner, inner, inner, inner, right],
  [bottomLeft, bottom, bottom, bottom, bottom, bottom, bottom, bottom, bottomRight],
  //
  [topLeft, top, top, top, top, top, top, top, topRight],
  [left, inner, inner, inner, inner, inner, inner, inner, right],
  [left, inner, inner, null, null, null, inner, inner, right],
  [left, inner, inner, null, null, null, inner, inner, right],
  [bottomLeft, bottom, bottom, null, null, null, bottom, bottom, bottomRight],
]);

const getImageByIndex = i => flatten(imageByIndex)[i];

export default getImageByIndex;
