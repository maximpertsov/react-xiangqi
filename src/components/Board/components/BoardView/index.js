import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { useSelector } from 'react-redux';

import {
  getBottomPlayerIsRed,
  getIsMoving,
  getSelectedMove,
  getTargets,
} from 'reducers';

import * as styles from 'commonStyles';
import { encode, moveToSquares } from 'services/logic/square';

import Square from './components/Square';
import XiangqiPiece from './components/Piece';
import LastMoveIndicator from './components/LastMoveIndicator';
import KingInCheckIndicator from './components/KingInCheckIndicator';
import SelectionIndicator from './components/SelectionIndicator';
import TargetIndicator from './components/TargetIndicator';

import boardImg from './assets/board-1000px.svg.png';

const Wrapper = styled.div`
  background-image: url(${boardImg});
  background-size: contain;
  background-repeat: no-repeat;
  background-position: top;
  display: grid;
  ${styles.MEDIA_TINY} {
    grid-template-rows: repeat(10, ${styles.SQUARE_SIZE_TINY});
    grid-template-columns: repeat(9, ${styles.SQUARE_SIZE_TINY});
  }
  ${styles.MEDIA_SMALL} {
    grid-template-rows: repeat(10, ${styles.SQUARE_SIZE_SMALL});
    grid-template-columns: repeat(9, ${styles.SQUARE_SIZE_SMALL});
  }
  ${styles.MEDIA_MEDIUM} {
    grid-template-rows: repeat(10, ${styles.SQUARE_SIZE_MEDIUM});
    grid-template-columns: repeat(9, ${styles.SQUARE_SIZE_MEDIUM});
  }
  ${styles.MEDIA_LARGE} {
    grid-template-rows: repeat(10, ${styles.SQUARE_SIZE_LARGE});
    grid-template-columns: repeat(9, ${styles.SQUARE_SIZE_LARGE});
  }
`;

const BoardView = ({ handleSquareClick }) => {
  const bottomPlayerIsRed = useSelector(state => getBottomPlayerIsRed(state));
  const selectedSquare = useSelector(state => state.selectedSquare);
  const targets = useSelector(state => getTargets(state));
  const { board, move: selectedMove, givesCheck } = useSelector(state =>
    getSelectedMove(state),
  );
  const isMoving = useSelector(state => getIsMoving(state));
  const [moveX, moveY] = useSelector(state => state.animationOffset);

  const getPieceCode = useCallback(
    square => board.getPiece(square) || undefined,
    [board],
  );

  const getSlot = useCallback(
    (slots, i) => (bottomPlayerIsRed ? i : slots.length - i - 1),
    [bottomPlayerIsRed],
  );

  const inLastMove = useCallback(
    square => {
      if (selectedMove === null) return false;

      return moveToSquares(selectedMove).includes(square);
    },
    [selectedMove],
  );

  const kingIsInCheck = useCallback(
    square => givesCheck && board.activeKing() === square,
    [board, givesCheck],
  );

  const isSelected = useCallback(
    square => !isMoving && selectedSquare === square,
    [isMoving, selectedSquare],
  );

  const isTargeted = useCallback(
    square => {
      if (isMoving) return false;

      return targets.some(move => moveToSquares(move)[1] === square);
    },
    [isMoving, targets],
  );

  const isOccupied = useCallback(square => getPieceCode(square) !== undefined, [
    getPieceCode,
  ]);

  const renderPiece = useCallback(
    square => (
      <XiangqiPiece
        code={getPieceCode(square)}
        moveX={selectedSquare === square ? moveX : 0}
        moveY={selectedSquare === square ? moveY : 0}
      />
    ),
    [getPieceCode, moveX, moveY, selectedSquare],
  );

  /* eslint-disable complexity */
  const renderSquares = () =>
    board.placement.map((_, i, slots) => {
      const slot = getSlot(slots, i);
      const square = encode(slot);
      return (
        <Square
          className="Square"
          key={square}
          handleClick={handleSquareClick(square)}
        >
          {isOccupied(square) && renderPiece(square)}
          {inLastMove(square) && <LastMoveIndicator />}
          {kingIsInCheck(square) && <KingInCheckIndicator />}
          {isSelected(square) && <SelectionIndicator />}
          {isTargeted(square) && (
            <TargetIndicator occupied={isOccupied(square)} />
          )}
        </Square>
      );
    });
  /* eslint-enable complexity */

  return <Wrapper className="BoardView">{renderSquares()}</Wrapper>;
};

BoardView.propTypes = {
  handleSquareClick: PropTypes.func.isRequired,
};

export default BoardView;
