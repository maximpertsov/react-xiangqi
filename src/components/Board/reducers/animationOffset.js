import * as logic from 'services/logic';

const setAnimationOffset = (state, { bottomPlayerIsRed, fromSlot, toSlot }) => {
  const [fromY, fromX] = logic.getRankFile(fromSlot);
  const [toY, toX] = logic.getRankFile(toSlot);

  return [
    bottomPlayerIsRed ? toX - fromX : fromX - toX,
    bottomPlayerIsRed ? toY - fromY : fromY - toY,
  ];
};

const animationOffset = (state = [0, 0], action) => {
  switch (action.type) {
    case 'set_animation_offset':
      return setAnimationOffset(state, action);
    case 'unset_animation_offset':
      return [0, 0];
    default:
      return state;
  }
};

export default animationOffset;

export const getIsMoving = state => state.some(dim => dim !== 0);
