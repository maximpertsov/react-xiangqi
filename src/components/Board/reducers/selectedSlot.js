const selectedSlot = (state = null, action) => {
  switch (action.type) {
    case 'set_selected_slot':
      return action.selectedSlot;
    default:
      return state;
  }
};

export default selectedSlot;
