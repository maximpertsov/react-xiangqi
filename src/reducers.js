import update from 'immutability-helper';
import { combineReducers } from 'redux';

import XiangqiBoard, * as logic from 'services/logic';
import { Color } from 'services/logic/constants';

// Initial State

// TODO: define in logic class
/* eslint-disable-next-line max-len */
const DEFAULT_FEN = 'rheakaehr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RHEAKAEHR';

const getInitialMovesState = (fen = DEFAULT_FEN) => ([
  {
    piece: undefined,
    fromPos: undefined,
    toPos: undefined,
    board: new XiangqiBoard({ fen }),
    pending: false,
  },
]);

const initialPlayers = [
  { name: undefined, color: Color.RED },
  { name: undefined, color: Color.BLACK },
];

const getInitialState = () => ({
  loading: true,
  moves: getInitialMovesState(),
  moveCount: 0,
  players: initialPlayers,
  selectedMoveIdx: 0,
  foobar: 0,
});

// Actions

const incrementMoveCount = (state) => (
  { ...state, moveCount: state.moveCount + 1 }
);

const setSelectedMove = (state, index) => (
  { ...state, selectedMoveIdx: index }
);

const setPreviousMove = (state) => (
  {
    ...state,
    selectedMoveIdx: Math.max(state.selectedMoveIdx - 1, 0),
  }
);

const setNextMove = (state) => (
  {
    ...state,
    selectedMoveIdx: Math.min(state.selectedMoveIdx + 1, state.moveCount),
  }
);

const selectLastMove = (state) => {
  const { moves } = state;
  return setSelectedMove(state, moves.length - 1);
};

const addMove = (state, board, fromSlot, toSlot, pending) => {
  const newMove = {
    fromPos: logic.getRankFile(fromSlot),
    toPos: logic.getRankFile(toSlot),
    piece: board.getPiece(fromSlot),
    board: board.move(fromSlot, toSlot),
    pending,
  };
  return selectLastMove(
    incrementMoveCount(
      update(state, { moves: { $push: [newMove] } }),
    ),
  );
};

const cancelMoves = (state) => {
  const { moves, selectedMoveIdx } = state;
  const confirmedMoves = moves.filter(({ pending }) => !pending);

  return {
    ...state,
    moves: confirmedMoves,
    moveCount: confirmedMoves.length,
    selectedMoveIdx: Math.min(selectedMoveIdx, confirmedMoves.length - 1),
  };
};

const confirmMoves = (state) => {
  const { moves } = state;
  return {
    ...state,
    moves: moves.map((move) => ({ ...move, pending: false })),
  };
};

const setMove = (
  state,
  { origin: fromPos, destination: toPos },
  fromMoveIdx = undefined,
) => {
  const { board } = state.moves[fromMoveIdx || state.moves.length - 1];
  const newMove = {
    piece: board.getPiece(fromPos, logic.RefType.RANK_FILE),
    fromPos,
    toPos,
    board: board.move(fromPos, toPos, logic.RefType.RANK_FILE),
    pending: false,
  };
  return incrementMoveCount(update(state, { moves: { $push: [newMove] } }));
};


const setMoves = (state, moves) => selectLastMove(
  moves.reduce(
    (prevState, move) => setMove(prevState, move),
    {
      ...state,
      moves: getInitialMovesState(),
      moveCount: 0,
      setSelectedMove: 0,
      loading: false,
    },
  ),
);

/* eslint-disable-next-line complexity */
const reducer = (state = getInitialState(), action) => {
  switch (action.type) {
  case 'increment_foobar':
    return ({ ...state, foobar: state.foobar + 1 });
  case 'add_move':
    return addMove(
      state,
      action.board,
      action.fromSlot,
      action.toSlot,
      action.pending,
    );
  case 'cancel_moves':
    return cancelMoves(state);
  case 'confirm_moves':
    return confirmMoves(state);
  case 'select_move':
    return setSelectedMove(state, action.index);
  case 'select_previous_move':
    return setPreviousMove(state);
  case 'select_next_move':
    return setNextMove(state);
  case 'set_moves':
    return setMoves(state, action.moves);
  case 'set_players':
    return ({ ...state, players: action.players });
  default:
    return state;
  }
};

export default combineReducers({ game: reducer });
