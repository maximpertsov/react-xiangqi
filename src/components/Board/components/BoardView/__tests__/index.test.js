import React from 'react';
import { useSelector } from 'react-redux';
import { getBottomPlayerIsRed } from 'reducers/selectors';

import BoardView from '..';

jest.mock('react-redux');
jest.mock('reducers/selectors');

describe('BoardView', () => {
  const store = mockStore({
    selectedFen:
      'rnbak1bnr/4a4/1c5c1/p1p1p1p1p/9/2P3P2/P3P3P/1C5C1/9/RNBAKABNR b - - 3 2',
  });

  let wrapper;

  beforeEach(() => {
    useSelector.mockImplementation(callback => callback(store.getState()));
    wrapper = shallowWrappedComponent(<BoardView />, store);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('bottom player has red pieces', () => {
    beforeAll(() => {
      getBottomPlayerIsRed.mockReturnValue(true);
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('bottom player has black pieces', () => {
    beforeAll(() => {
      getBottomPlayerIsRed.mockReturnValue(false);
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
