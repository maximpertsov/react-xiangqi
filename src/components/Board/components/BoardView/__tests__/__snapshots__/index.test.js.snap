// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardView bottom player has black pieces snapshot 1`] = `
<Styled(div)
  className="BoardView"
>
  <Square
    key="i1"
    square="i1"
  />
  <Square
    key="h1"
    square="h1"
  />
  <Square
    key="g1"
    square="g1"
  />
  <Square
    key="f1"
    square="f1"
  />
  <Square
    key="e1"
    square="e1"
  />
  <Square
    key="d1"
    square="d1"
  />
  <Square
    key="c1"
    square="c1"
  />
  <Square
    key="b1"
    square="b1"
  />
  <Square
    key="a1"
    square="a1"
  />
  <Square
    key="i2"
    square="i2"
  />
  <Square
    key="h2"
    square="h2"
  />
  <Square
    key="g2"
    square="g2"
  />
  <Square
    key="f2"
    square="f2"
  />
  <Square
    key="e2"
    square="e2"
  />
  <Square
    key="d2"
    square="d2"
  />
  <Square
    key="c2"
    square="c2"
  />
  <Square
    key="b2"
    square="b2"
  />
  <Square
    key="a2"
    square="a2"
  />
  <Square
    key="i3"
    square="i3"
  />
  <Square
    key="h3"
    square="h3"
  />
  <Square
    key="g3"
    square="g3"
  />
  <Square
    key="f3"
    square="f3"
  />
  <Square
    key="e3"
    square="e3"
  />
  <Square
    key="d3"
    square="d3"
  />
  <Square
    key="c3"
    square="c3"
  />
  <Square
    key="b3"
    square="b3"
  />
  <Square
    key="a3"
    square="a3"
  />
  <Square
    key="i4"
    square="i4"
  />
  <Square
    key="h4"
    square="h4"
  />
  <Square
    key="g4"
    square="g4"
  />
  <Square
    key="f4"
    square="f4"
  />
  <Square
    key="e4"
    square="e4"
  />
  <Square
    key="d4"
    square="d4"
  />
  <Square
    key="c4"
    square="c4"
  />
  <Square
    key="b4"
    square="b4"
  />
  <Square
    key="a4"
    square="a4"
  />
  <Square
    key="i5"
    square="i5"
  />
  <Square
    key="h5"
    square="h5"
  />
  <Square
    key="g5"
    square="g5"
  />
  <Square
    key="f5"
    square="f5"
  />
  <Square
    key="e5"
    square="e5"
  />
  <Square
    key="d5"
    square="d5"
  />
  <Square
    key="c5"
    square="c5"
  />
  <Square
    key="b5"
    square="b5"
  />
  <Square
    key="a5"
    square="a5"
  />
  <Square
    key="i6"
    square="i6"
  />
  <Square
    key="h6"
    square="h6"
  />
  <Square
    key="g6"
    square="g6"
  />
  <Square
    key="f6"
    square="f6"
  />
  <Square
    key="e6"
    square="e6"
  />
  <Square
    key="d6"
    square="d6"
  />
  <Square
    key="c6"
    square="c6"
  />
  <Square
    key="b6"
    square="b6"
  />
  <Square
    key="a6"
    square="a6"
  />
  <Square
    key="i7"
    square="i7"
  />
  <Square
    key="h7"
    square="h7"
  />
  <Square
    key="g7"
    square="g7"
  />
  <Square
    key="f7"
    square="f7"
  />
  <Square
    key="e7"
    square="e7"
  />
  <Square
    key="d7"
    square="d7"
  />
  <Square
    key="c7"
    square="c7"
  />
  <Square
    key="b7"
    square="b7"
  />
  <Square
    key="a7"
    square="a7"
  />
  <Square
    key="i8"
    square="i8"
  />
  <Square
    key="h8"
    square="h8"
  />
  <Square
    key="g8"
    square="g8"
  />
  <Square
    key="f8"
    square="f8"
  />
  <Square
    key="e8"
    square="e8"
  />
  <Square
    key="d8"
    square="d8"
  />
  <Square
    key="c8"
    square="c8"
  />
  <Square
    key="b8"
    square="b8"
  />
  <Square
    key="a8"
    square="a8"
  />
  <Square
    key="i9"
    square="i9"
  />
  <Square
    key="h9"
    square="h9"
  />
  <Square
    key="g9"
    square="g9"
  />
  <Square
    key="f9"
    square="f9"
  />
  <Square
    key="e9"
    square="e9"
  />
  <Square
    key="d9"
    square="d9"
  />
  <Square
    key="c9"
    square="c9"
  />
  <Square
    key="b9"
    square="b9"
  />
  <Square
    key="a9"
    square="a9"
  />
  <Square
    key="i10"
    square="i10"
  />
  <Square
    key="h10"
    square="h10"
  />
  <Square
    key="g10"
    square="g10"
  />
  <Square
    key="f10"
    square="f10"
  />
  <Square
    key="e10"
    square="e10"
  />
  <Square
    key="d10"
    square="d10"
  />
  <Square
    key="c10"
    square="c10"
  />
  <Square
    key="b10"
    square="b10"
  />
  <Square
    key="a10"
    square="a10"
  />
</Styled(div)>
`;

exports[`BoardView bottom player has red pieces snapshot 1`] = `
<Styled(div)
  className="BoardView"
>
  <Square
    key="a10"
    square="a10"
  />
  <Square
    key="b10"
    square="b10"
  />
  <Square
    key="c10"
    square="c10"
  />
  <Square
    key="d10"
    square="d10"
  />
  <Square
    key="e10"
    square="e10"
  />
  <Square
    key="f10"
    square="f10"
  />
  <Square
    key="g10"
    square="g10"
  />
  <Square
    key="h10"
    square="h10"
  />
  <Square
    key="i10"
    square="i10"
  />
  <Square
    key="a9"
    square="a9"
  />
  <Square
    key="b9"
    square="b9"
  />
  <Square
    key="c9"
    square="c9"
  />
  <Square
    key="d9"
    square="d9"
  />
  <Square
    key="e9"
    square="e9"
  />
  <Square
    key="f9"
    square="f9"
  />
  <Square
    key="g9"
    square="g9"
  />
  <Square
    key="h9"
    square="h9"
  />
  <Square
    key="i9"
    square="i9"
  />
  <Square
    key="a8"
    square="a8"
  />
  <Square
    key="b8"
    square="b8"
  />
  <Square
    key="c8"
    square="c8"
  />
  <Square
    key="d8"
    square="d8"
  />
  <Square
    key="e8"
    square="e8"
  />
  <Square
    key="f8"
    square="f8"
  />
  <Square
    key="g8"
    square="g8"
  />
  <Square
    key="h8"
    square="h8"
  />
  <Square
    key="i8"
    square="i8"
  />
  <Square
    key="a7"
    square="a7"
  />
  <Square
    key="b7"
    square="b7"
  />
  <Square
    key="c7"
    square="c7"
  />
  <Square
    key="d7"
    square="d7"
  />
  <Square
    key="e7"
    square="e7"
  />
  <Square
    key="f7"
    square="f7"
  />
  <Square
    key="g7"
    square="g7"
  />
  <Square
    key="h7"
    square="h7"
  />
  <Square
    key="i7"
    square="i7"
  />
  <Square
    key="a6"
    square="a6"
  />
  <Square
    key="b6"
    square="b6"
  />
  <Square
    key="c6"
    square="c6"
  />
  <Square
    key="d6"
    square="d6"
  />
  <Square
    key="e6"
    square="e6"
  />
  <Square
    key="f6"
    square="f6"
  />
  <Square
    key="g6"
    square="g6"
  />
  <Square
    key="h6"
    square="h6"
  />
  <Square
    key="i6"
    square="i6"
  />
  <Square
    key="a5"
    square="a5"
  />
  <Square
    key="b5"
    square="b5"
  />
  <Square
    key="c5"
    square="c5"
  />
  <Square
    key="d5"
    square="d5"
  />
  <Square
    key="e5"
    square="e5"
  />
  <Square
    key="f5"
    square="f5"
  />
  <Square
    key="g5"
    square="g5"
  />
  <Square
    key="h5"
    square="h5"
  />
  <Square
    key="i5"
    square="i5"
  />
  <Square
    key="a4"
    square="a4"
  />
  <Square
    key="b4"
    square="b4"
  />
  <Square
    key="c4"
    square="c4"
  />
  <Square
    key="d4"
    square="d4"
  />
  <Square
    key="e4"
    square="e4"
  />
  <Square
    key="f4"
    square="f4"
  />
  <Square
    key="g4"
    square="g4"
  />
  <Square
    key="h4"
    square="h4"
  />
  <Square
    key="i4"
    square="i4"
  />
  <Square
    key="a3"
    square="a3"
  />
  <Square
    key="b3"
    square="b3"
  />
  <Square
    key="c3"
    square="c3"
  />
  <Square
    key="d3"
    square="d3"
  />
  <Square
    key="e3"
    square="e3"
  />
  <Square
    key="f3"
    square="f3"
  />
  <Square
    key="g3"
    square="g3"
  />
  <Square
    key="h3"
    square="h3"
  />
  <Square
    key="i3"
    square="i3"
  />
  <Square
    key="a2"
    square="a2"
  />
  <Square
    key="b2"
    square="b2"
  />
  <Square
    key="c2"
    square="c2"
  />
  <Square
    key="d2"
    square="d2"
  />
  <Square
    key="e2"
    square="e2"
  />
  <Square
    key="f2"
    square="f2"
  />
  <Square
    key="g2"
    square="g2"
  />
  <Square
    key="h2"
    square="h2"
  />
  <Square
    key="i2"
    square="i2"
  />
  <Square
    key="a1"
    square="a1"
  />
  <Square
    key="b1"
    square="b1"
  />
  <Square
    key="c1"
    square="c1"
  />
  <Square
    key="d1"
    square="d1"
  />
  <Square
    key="e1"
    square="e1"
  />
  <Square
    key="f1"
    square="f1"
  />
  <Square
    key="g1"
    square="g1"
  />
  <Square
    key="h1"
    square="h1"
  />
  <Square
    key="i1"
    square="i1"
  />
</Styled(div)>
`;
